kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,426 @@
+///       In-commit timestamps is enabled.
+///     - Ensure that reading is supported on any protocol updates
+///     - Ensure that Change Data Feed is enabled for any metadata update. See  [`TableProperties`]
+///     - Ensure that any schema update is compatible with the provided `schema`. Currently, schema
+///       compatibility is checked through schema equality. This will be expanded in the future to
+///       allow limited schema evolution.
+///     - Determine if there exist any `cdc` actions.
+///     - Constructs the remove deletion vector map from paths belonging to `remove` actions to the
+///       action's corresponding [`DvInfo`]. This map will be filtered to only contain paths that
+///       exists in another `add` action _within the same commit_. This corresponds to `remove_dv`
+///       in [`TableChangesScanData`].
+/// 2. Scan file generation phase [`LogReplayScanner::into_scan_batches`]: This performs another
+///    iteration over every action in the commit, and generates [`TableChangesScanData`]. It does
+///    so by transforming the actions using [`get_add_transform_expr`], and generating selection
+///    vectors with the following rules:
+///     - If a `cdc` action was found in the prepare phase, only `cdc` actions are selected
+///     - Otherwise, select `add` and `remove` actions. Note that only `remove` actions that do not
+///       share a path with an `add` action are selected.
+///
+/// Note: As a consequence of the two phases, LogReplayScanner will iterate over each action in the
+/// commit twice. It also may use an unbounded amount of memory, proportional to the number of
+/// `add` + `remove` actions in the _single_ commit.
+struct LogReplayScanner {
+    // True if a `cdc` action was found after running [`LogReplayScanner::prepare_table_changes`]
+    has_cdc_action: bool,
+    // A map from path to the deletion vector from the remove action. It is guaranteed that there
+    // is an add action with the same path in this commit
+    remove_dvs: HashMap<String, DvInfo>,
+    // The commit file that this replay scanner will operate on.
+    commit_file: ParsedLogPath,
+    // The timestamp associated with this commit. By default this is the file modification time
+    // from the commit's [`FileMeta`]. If there is a [`CommitInfo`] with a timestamp generated by
+    // in-commit timestamps, that timestamp will be used instead.
+    //
+    // Note: This will be used once an expression is introduced to transform the engine data in
+    // [`TableChangesScanData`]
+    #[allow(unused)]
+    timestamp: i64,
+    // The data skipping filter for filtering log actions
+    filter: Option<DataSkippingFilter>,
+}
+
+fn prepare_table_changes(
+    commit_file: ParsedLogPath,
+    engine: &dyn Engine,
+    table_schema: &SchemaRef,
+    physical_predicate: Option<ExpressionRef>,
+) -> DeltaResult<LogReplayScanner> {
+    let filter = DataSkippingFilter::new(engine, table_schema, physical_predicate);
```

Comment:
iirc @nicklan  wanted the filter inside since it makes it easier to expose this function to FFI. I can go ahead and move the filter construction out, but just wanted to flag that.

Reply:
I don't think we should be exposing this to the engine in the first place? Shouldn't the "public" API be that engine passes in the raw inputs (including an expression to be used for data skipping), and the result is an iterator that can be visited? Or even an internal iterator method that takes a callback from engine?

Reply:
True, the engine would likely be interacting with `TableChangesScan::scan_data`